.github/workflows/benchmark.yml

```diff
@@ -0,0 +1,156 @@
+  push:
+    branches: [main]
+
+permissions:
+  contents: read
+
+defaults:
+  run:
+    shell: bash
+#TODO  concurrency groups?
+jobs:
+  benchmark:
+    runs-on: 8-core
+    container: ghcr.io/facebookincubator/velox-dev:amd64-ubuntu-22.04-avx
+    env:
+      CCACHE_DIR: "${{ github.workspace }}/.ccache/"
+      CCACHE_BASEDIR: "${{ github.workspace }}"
+      BINARY_DIR: "${{ github.workspace }}/benchmarks/"
+      CONBENCH_URL: "https://velox-conbench.voltrondata.run/"
+      CONBENCH_HOST: "GitHub-runner"
+      LINUX_DISTRO: "ubuntu"
+      RESULTS_ROOT: "${{ github.workspace }}/benchmark-results"
+      BASELINE_OUTPUT_PATH: "${{ github.workspace }}/benchmark-results/baseline/"
+      TARGET_OUTPUT_PATH: "${{ github.workspace }}/benchmark-results/target/"
+    steps:
+      - name: "Setup ccache and python"
+        run: |
+            # Set up ccache configs .
+            mkdir -p .ccache
+            ccache -sz -M 5Gi
+            
+            # We can not use setup-python as we are running in a container
+            # TODO add to docker image and remove here
+            apt update
+            apt install -y lsb-release python3 pip
+
+      - uses: actions/cache/restore@v3
+        id: restore-cache
+        with:
+          path: ".ccache"
+          key: ccache-benchmark-${{ github.sha }}
+          restore-keys: |
+            ccache-benchmark-
+
+      - name: "Checkout Base"
+        uses: actions/checkout@v3
+        with:
+          path: 'velox'
+          ref: ${{ github.event.pull_request.base.sha || github.sha }}
+          submodules: 'recursive'
+
+      - name: Build Base Benchmarks
+        working-directory: velox
+        run: |
+            n_cores=$(nproc)
+            make benchmarks-basic-build NUM_THREADS=$n_cores MAX_HIGH_MEM_JOBS=$((n_cores/2)) MAX_LINK_JOBS=$n_cores
+            ccache -s
+            mkdir -p ${BINARY_DIR}/baseline/
+            cp -r --verbose _build/release/velox/benchmarks/basic/* ${BINARY_DIR}/baseline/
+
+      - name: "Checkout Head"
+        if: ${{ github.event_name == 'pull_request' }}
+        uses: actions/checkout@v3
+        with:
+          path: 'velox'
+          submodules: 'recursive'
+
+      - name: Build Target Benchmarks
+        if: ${{ github.event_name == 'pull_request' }}
+        working-directory: velox
+        run: |
+            n_cores=$(nproc)
+            make benchmarks-basic-build NUM_THREADS=$n_cores MAX_HIGH_MEM_JOBS=$n_cores MAX_LINK_JOBS=$n_cores
+            ccache -s
+            mkdir -p  ${BINARY_DIR}/target/
+            cp -r --verbose _build/release/velox/benchmarks/basic/*  ${BINARY_DIR}/target/
+
+      - uses: actions/cache/save@v3
+        id: cache
+        with:
+          path: ".ccache"
+          key: ccache-benchmark-${{ github.sha }}
+
+      - uses: actions/upload-artifact@v3
+        with:
+          path: ${{ env.BINARY_DIR }}
+          name: "benchmark-binaries"
+
+      - name: Remove Unchanged Benchmarks
```

Comment:
I think I'd rather we never do this step. From a conbench standpoint, it's valuable to get as many benchmark measurements on one state of code as possible, so that we can be more certain what the acceptable distribution looks like. So I think it's worth the build time to always run all benchmarks on a new commit, even if some technically haven't changed.

Reply:
Question: does this step remove the source files of the benchmarks (if the source is unchanged) because we can simply use the cached version of that same benchmark file from the cache? In which case we always _run_ every benchmark each time — but we might not _compile_ each benchmark file each time? Or does it remove them from running as well?

If we are skipping the running, I strongly agree with Austin: we shouldn't do that we should run (and send to conbench) all benchmarks we can so that we have a broader sample to measure history, stability, and regression from

Reply:
Oh yeah, I was typing under the assumption that this step skips the _running_ of benchmarks whose source didn't change.

Reply:
No it removes both versions of duplicate executables to not run them. I am fine with running them anyway if you prefer that. 

This makes sense:
> as many benchmark measurements on one state of code as possible

but the way we do it for unchanged benchmarks the additional run is associated with the contender SHA instead of the base SHA. Should it not be saved as a run for the base SHA? But detecting that reliably (I am not sure the above code actually works properly, as I have not tested it with known actually distinct/changed benchmarks) might be more trouble than it's worth? 

We could cache the binaries for base and only build on a cache miss. That might speed up the build a bit.

Reply:
I am curious @assignUser  even if the source doesnt change are the executables guaranteed to be byte for byte same - I am not sure if the compilation/linking process always produces completely identical binaries.

Reply:
As we build both version in the same folder they should not be rebuilt if they are unaffected by the change but as stated above I am not 100% on this which is why I deactivated this for now and will do some additional testing later on as I see this as a bonus feature that would be nice to have but is not crucial for the workflow to be merged,

Reply:
>Should it not be saved as a run for the base SHA?

IMO we shouldn't ever upload the `baseline` results to Conbench. Running `baseline` benchmarks is a semi-temporary thing. In the future we will only be running/uploading `target` results, and comparing them to the baseline results that already live in Conbench, and maybe re-running based on that.